File: high_perf_science_compute_keller/interpolation/julia/linear_interpolation.jl
using Printf
const BASE_RES = 10

mutable struct Result
    xs::Array{Float64}
    ys::Array{Float64}
end

function interp(y0, y1, x0, x1, x)
    y0 + (x - x0) * (y1 - y0) / (x1 - x0)
end

function interpolate(xFit, yFit, n)
    scap = 10 * (n - 1) #90
    xs = Array{Float64}(undef, scap)
    ys = Array{Float64}(undef, scap)
    for i = 1:(n - 1)
        icount = 0
        for j = 1:10
            delta = xFit[i + 1] - xFit[i]
            idx = (10 * icount) + j
            xs[idx] = xFit[i] + 0.1 * float(j) * delta
            ys[idx] = interp(yFit[i], yFit[i + 1], xFit[i], xFit[i + 1], xs[idx])
        end
        icount += 1
    end
    Result(xs, ys)
end

function run()
    xsi = Array{Float64}(undef, BASE_RES)
    ysi = Array{Float64}(undef, BASE_RES)

    for i = 1:BASE_RES
        xsi[i] = 4.0 * π * float(i) / float(BASE_RES)
        ysi[i] = sin(xsi[i])
    end

    res = interpolate(xsi, ysi, BASE_RES)

    
    for i = 1:(10 * (BASE_RES - 1))
        @printf("%e %e\n", res.xs[i], res.ys[i])
    end

end

#not correct
#julia linear_interpolation.jl > linplot_out
# ./plot_interp.py julia/linplot_out
run()

File: high_perf_science_compute_keller/interpolation/c/linear_interpolation.c
#include "math.h"
#include "stdio.h"
#include "stdlib.h"

// the number of points in original data set
#define BASE_RES 10

//this struct will contain the output of interpolation process
typedef struct result
{
    float *xs; //array of floats for x points
    float *ys; //array of floats for y points
} result;

// given two sets of (x,y) points and a new x val,
// return the linear interpolated y value
// y = mx+b   == rise / run
// rise == (y1 - y0)
// run == (x1 - x0)
// we use an offset y0 + (x - x0) since we trying to find value as this x point
float interp(float y0, float y1, float x0, float x1, float x)
{
    //first lagrange polynomial
    return y0 + (x - x0) * (y1 - y0) / (x1 - x0);
}

result interpolate(float *xs, float *ys, int n)
{
    //unsigned long
    //printf("malloc: %lu\n\n", (10 * (n - 1) * sizeof(float)));
    //printf("sizeof(float): %lu\n\n", sizeof(float));
    float delta;
    result res;
    res.xs = malloc(10 * (n - 1) * sizeof(float));
    res.ys = malloc(10 * (n - 1) * sizeof(float));
    for (int i = 0; i < n - 1; i++) //loop n-1 times
    {
        for (int j = 0; j < 10; j++)
        {
            delta = xs[i + 1] - xs[i];
            //printf("   i:%d, j:%d;   10 * i + j == %d\n", i, j, 10 * i + j);
            res.xs[10 * i + j] = xs[i] + 0.1 * ((float)j) * delta;
            // interpolate 10 points between x-y
            res.ys[10 * i + j] = interp(ys[i], ys[i + 1], xs[i], xs[i + 1], res.xs[10 * i + j]);
        }
    }
    return res;
}

/*
 *  gcc linear_interpolation.c -lm -o linear
 *  ./linear > linplot_out
 *  ./plot_interp.py c/linplot_out
 */
int main()
{
    //generate input data (simple sine curve)
    float xs[BASE_RES];
    float ys[BASE_RES];
    for (int i = 0; i < BASE_RES; i++)
    {
        xs[i] = 4. * M_PI * ((float)i) / BASE_RES;
        ys[i] = sin(xs[i]);
    }

    result res = interpolate(xs, ys, BASE_RES);
    //interpolate to 10x higher resolution
    for (int i = 0; i < 10 * (BASE_RES - 1); i++)
    {
        printf("%e %e\n", res.xs[i], res.ys[i]);
    }
    return 0;
}
